### PR TITLE
Migrates leader election to Lease API

### LINKED ISSUE #111: Migrate leader election to use Lease API
Current leader election is using endpoints

https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/853cdd7cba6e20753071cdb3e07fe37a6f2fc5a3/controller/controller.go#L867

However, that has some problems and is recommended to use the Lease API instead

https://github.com/kubernetes/kubernetes/issues/80289

### PR DESCRIPTION
**Pull request by snOm3ad**
This PR migrates the leader election to the Lease API.

I ran `make test` and got no errors back, I also ran the `tests/e2e/test.sh` script 
and the process completed successfully. I believe this change completes the migration 
to the Lease API but I'm not entirely certain. If someone can verify this and/or provide 
any other feedback I'll greatly appreciate it!

Please see #111 and this ongoing [issue](https://github.com/kubernetes/kubernetes/issues/80289) for more information.

Fixes #111

### CODE REVIEW

**Comment by jsafrane** [comment]
@snOm3ad As you can see in https://github.com/kubernetes/kubernetes/issues/80289, moving from endpoints to leases needs to be two step process. To make sure an old provisioner that uses endpoints is not running, the new provisioner needs to acquire both endpoints-based and lease-bases leases. Only after gratuitous time, when we can assume that no provisioner in a cluster could ever use endpoints-based lease, a new provisioner can use only lease-based lease.

In other words, the leader election should be configurable in the library. A caller should be able to configure if they want endpoints (default), leases or both.

**Comment by snOm3ad** [comment]
I see, thanks for the explanation! It seems that is the general pattern whenever you want to change something in a cluster, very similar to how a rolling update works.

File: controller/controller.go [hunk 1]
```diff
@@ -864,11 +864,11 @@ func (ctrl *ProvisionController) Run(ctx context.Context) {
 	go ctrl.volumeStore.Run(ctx, DefaultThreadiness)
 
 	if ctrl.leaderElection {
-		rl, err := resourcelock.New("endpoints",
+		rl, err := resourcelock.New("leases",
 			ctrl.leaderElectionNamespace,
 			strings.Replace(ctrl.provisionerName, "/", "-", -1),
-			ctrl.client.CoreV1(),
 			nil,
+			ctrl.client.CoordinationV1(),
```
Comment (NikhilSharmaWe) [inline controller/controller.go:871]: @snOm3ad Could you please make me understand what was the reason for changing this argument.
